core/tx_pool.go

```diff
@@ -640,6 +640,10 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (bool, error) {
 		pool.journalTx(from, tx)
 
 		log.Trace("Pooled new executable transaction", "hash", hash, "from", from, "to", tx.To())
+
+		// We've directly injected a replacement transaction, notify subsystems
+		go pool.txFeed.Send(TxPreEvent{tx})
```

Comment:
The code indicates that `old` can be `nil` - what scenario would make that happen? And if it is, is it still correct to fire this event?

Reply:
Good question. Old should not be able to be `nil` here if we're checking for nonce overlaps in the outer `if`. That being said, I wouldn't feel comfortable removing the `nil` check as it's a nice sanity protection feature. It also permits perhaps `list.Overlaps(tx)` to return true if the new transaction is *extending* the currently pending list, not *overwriting* part of it (though that's a whole new PR, not a small fix).

As for the negative effect, I don't see any. The transaction announcement event is used by the miner (if not mining, to potentially update the pending state) and the `eth` handler to propagate to peers (but it dedups all transactions to avoid network junk).